vertical-pod-autoscaler/enhancements/7397-max-allowed-recommendation/README.md

```diff
@@ -0,0 +1,45 @@
+# KEP-7397: Max allowed recommendation configurable
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+   - [Goals](#goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+   - [Test Plan](#test-plan)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+VPA currently allows configuring the maximum allowed recommendation per resource using the field `.spec.resourcePolicy.containerPolicies[].maxAllowed`. This helps ensure that recommendations don't exceed the allocatable capacity of the largest node. Without this limit, a high recommendation could make the Pod unschedulable, causing downtime. We aim to improve the VPA recommender by making it aware of the maximum node size, preventing it from generating recommendations that would render Pods unschedulable.
+
+## Motivation
+In dynamic environments, manual configuration of .maxAllowed can be error-prone, especially as cluster sizes and node configurations change over time. Automating this process by making the VPA recommender aware of the largest node size ensures more reliable scaling decisions. This helps prevents situations where Pods become unschedulable due to excessive resource recommendations.
+
+### Goals
+* Make Recommendations More Accurate and Easier to Schedule: Try to have the VPA recommender suggest resources that fit within the largest node’s capacity, helping pods get scheduled more easily and lowering the chance of downtime.
+* Adapt to Changing Environments: Adjust recommendations as the cluster size or node configurations change to better reflect the current state of the cluster.
+
+## Proposal
+VPA could query the Cluster Autoscaler (CA) for feedback, potentially using the [provisioning-request](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/provisioning-request.md) API or a similar mechanism. This approach aligns well with CA’s design, as it is built to handle questions related to resource availability and scheduling. However, this solution comes with some challenges:  
+
+- Not all clusters use CA, and Karpenter currently does not support the `provisioningrequest` API and is unlikely to support it in the future.  
```

Comment:
As long as the behavior is clearly defined behind a flag (meaning it's clear to the user whether they can rely on CA) I'd say it's fine.
Admission would need to check if the API is available and prevent people from relying on the feature if it's not.

I also wonder about VPA reaction time with this approach. Would VPA be able to cache available node types?

Reply:
Yes, this behavior will be controlled via a flag, maybe indicating whether it's using Karpenter or CA.
AFAIK, The VPA should be able to cache the available node types. 
wdyt? @adrianmoisey

Reply:
Caching seems sane to me. I agree, that the round-trip time from the admission-controller, to CA and back again may be slow, so caching may be needed.
And I agree, this setting needs to be opt-in only. And it should warn the user when the cluster-autoscaler is not present.

Reply:
Ill add it to the KEP then